File: experiment-5/runs/language=java_model=claude-opus-4-8_tooling=none/rep1/src/main/java/com/brasilsoccer/mcp/model/Match.java
/*
 * ============================================================================
 * Match - immutable record of a single soccer match
 * ============================================================================
 * Context:
 *   Unified representation of a match coming from any of the five match CSV
 *   datasets (Brasileirao, Copa do Brasil, Libertadores, BR-Football extended
 *   stats, and the historical 2003-2019 Brasileirao). Each source has slightly
 *   different columns; the loader normalises them all into this shape.
 *
 *   Team names are stored in three forms (see TeamNames):
 *     - homeTeam   / awayTeam    : faithful display name, state suffix kept.
 *     - homeKey    / awayKey     : base key (suffix stripped) for cross-dataset
 *       matching, so "Palmeiras-SP", "Palmeiras" and "palmeiras" compare equal.
 *     - homeFullKey/ awayFullKey : suffix-qualified key that keeps same-base
 *       clubs distinct ("atletico-mg" vs "atletico-pr") for standings.
 *
 *   `competition` is the canonical competition name and `source` identifies the
 *   originating dataset, which lets standings logic pick one authoritative
 *   source per (competition, season) and avoid double-counting overlapping data.
 * ============================================================================
 */
package com.brasilsoccer.mcp.model;

import java.time.LocalDate;
import java.util.Optional;

public record Match(
        String competition,
        String source,
        LocalDate date,
        Integer season,
        String round,
        String stage,
        String stadium,
        String homeTeam,
        String awayTeam,
        String homeKey,
        String awayKey,
        String homeFullKey,
        String awayFullKey,
        int homeGoal,
        int awayGoal) {

    /** Total goals in the match. */
    public int totalGoals() {
        return homeGoal + awayGoal;
    }

    /** Absolute goal difference (margin of victory). */
    public int margin() {
        return Math.abs(homeGoal - awayGoal);
    }

    public boolean isDraw() {
        return homeGoal == awayGoal;
    }

    public boolean homeWin() {
        return homeGoal > awayGoal;
    }

    public boolean awayWin() {
        return awayGoal > homeGoal;
    }

    /** True if the given canonical team key played in this match. */
    public boolean involves(String teamKey) {
        return homeKey.equals(teamKey) || awayKey.equals(teamKey);
    }

    public Optional<LocalDate> dateOpt() {
        return Optional.ofNullable(date);
    }

    /**
     * Stable de-duplication key. Two records from different source datasets that
     * describe the same fixture (same teams, date and score) collapse to one.
     */
    public String dedupeKey() {
        return homeKey + "|" + awayKey + "|" + (date == null ? "?" : date)
                + "|" + homeGoal + "-" + awayGoal;
    }
}

File: experiments/adrianco/experiment-48-fable5-gaps/brazil/runs/language=elixir_model=claude-fable-5_prompt=neutral/rep1/lib/brazilian_soccer_mcp/match.ex
defmodule BrazilianSoccerMcp.Match do
  @moduledoc """
  A single match, normalized from any of the five match CSV files.

    * `:source`      - which file the record came from
                       (`:serie_a`, `:historical`, `:copa`, `:libertadores`, `:extended`)
    * `:competition` - `:brasileirao`, `:serie_b`, `:serie_c`, `:copa_do_brasil`,
                       `:libertadores`
    * `:home_key`/`:away_key` - canonical team keys (see `TeamNames`)
    * `:extras`      - source-specific detail (stadium, corners, shots, ...)
  """

  @enforce_keys [:source, :competition, :date, :home_key, :away_key]
  defstruct [
    :source,
    :competition,
    :date,
    :time,
    :season,
    :round,
    :stage,
    :home,
    :away,
    :home_key,
    :away_key,
    :home_goal,
    :away_goal,
    extras: %{}
  ]

  def winner_key(%__MODULE__{home_goal: hg, away_goal: ag}) when is_nil(hg) or is_nil(ag), do: nil

  def winner_key(%__MODULE__{home_goal: hg, away_goal: ag} = m) do
    cond do
      hg > ag -> m.home_key
      ag > hg -> m.away_key
      true -> :draw
    end
  end

  def margin(%__MODULE__{home_goal: hg, away_goal: ag}) when is_nil(hg) or is_nil(ag), do: 0
  def margin(%__MODULE__{home_goal: hg, away_goal: ag}), do: abs(hg - ag)
end
